doc/guides/modules.rst

```diff
@@ -10,14 +10,342 @@ system those are called *modules*. These modules must be integrated with the
 this page.
 
-Zephyr depends on several categories of modules, including:
+To be classified as a candidate for being included in the default list of
+modules, an external project is required to have its own life-cycle outside
+the Zephyr Project, that is, reside in its own repository, and have its own
+contribution and maintainance workfow and release process. Zephyr modules
+should not contain code that is written exclusively for Zephyr.
+
+  .. note::
+
+     Code which is written for Zephyr and works exclusively with Zephyr
+     should not be submitted as a Zephyr module; instead, it should be
+     part of the main Zephyr tree.
+
+Modules to be included in the default manifest of the Zephyr project need to
+provide functionality or features endorsed and approved by the project technical
+steering committee and should comply with the
+:ref:`module licensing requirements<modules_licensing>` and
+:ref:`contribute guidelines<modules_contributing>`. They should also have a
+Zephyr developer that is commited to maintain the module code base.
+
+Zephyr depends on several categories of modules, including but not limited to:
 
 - Debugger integration
 - Silicon vendor Hardware Abstraction Layers (HALs)
 - Cryptography libraries
 - File Systems
 - Inter-Process Communication (IPC) libraries
 
+This page summarizes a list of policies and best practices which aim at
+better organizing the workflow in Zephyr modules.
+
+Module Repositories
+*******************
+
+* All modules included in the default manifest shall be hosted in repositories
+  under the Zephyr Project Github organization.
+
+* Module repository names should follow the convention of using lowercase
+  letters and dashes instead of underscores. This rule will apply to all
+  new module repositories.
+
+  .. note::
+
+     It is not required to modify the names of existing module repositories
+     to make them comply with the above naming convention.
+
+* If the module has an external (upstream) project repository, the module
+  repository should preserve the upstream repository folder structure.
+
+* Modules should use "zephyr" as the default name for the repository main
+  branch. Branches for specific purposes, for example, a module branch for
+  an LTS Zephyr version, shall have names starting with the 'zephyr\_' prefix.
+
+  .. note::
+
+     It is **not** recommended in module repositories to have a 'master'
+     branch mirroring the master branch of the external repository.
+
+.. _modules_synchronization:
+
+Synchronizing with upstream
+===========================
+
+Requirements for allowed practices
+----------------------------------
+
+Any changes to module repositories, as a result of synchronizing with,
+upstream shall be brought in by pull requests. These pull requests
+shall be merg-able (i.e. with Rebase & Merge option, using Github UI).
+This ensures that a linear history is preserved in the module repositories.
```

Comment:
@pabigot this will be discussed today, for a last iteration, and I'll give a note here.

Reply:
@pabigot 
I will add a sentence to justify this. In brief
- any changes to module repositories, need to be done with PRs. Here, we talk about synchronization only (that's because "normal" changes, e.g. bug fixes, are obviously brought in with PRs). The PRs shall be
    - mergable with Github's UI, which ensures clarity and transparency
    - verifiable via CI (CI triggers on the main repo with manifest pointing to module PR)
    - reviewed and approved properly.
Merge-able PRs ensure a proper level of visibility of what has changed, i.e. there's an easy (or less easy) way to get the list of commits between the old SHA and the new SHA in the .west manifest.

We must always be able to see, using GH's UI what has changed in between two manifest SHAs. This is a hard requirements agreed in the Process forum.

> I certainly hope that's consistent with the intent, because requiring us to do a rebase of the upstream changes onto the Zephyr head in order to update to a new version is a non-starter.

There is always the option of a simple single-commit diff; the "snapshot" solution, as a fall-back alternative.
We are also discussing a "merge"-based option where you still merge using GH UI.

Reply:
OK, there was probably confusion there, because of this "linear history" thing.  There's nothing about github's UI that requires the absence of merge commits, with the material being merged being the upstream commits in their original form (SHA1).  In practice, I expect the merges in the Zephyr repo will discard that information, though, because they'll be rebased.

Reply:
Agree, this "linear" history term was confusing. What I wanted to convey is that we need linear, incremental downstream history, e.g. no re-writes, (no force-pushing), which would make SHAs disappear. 

> There's nothing about github's UI that requires the absence of merge commits

Yes, and in fact, merge commits will appear, anyways, if the upstream project allows merge commits.

Allowing downstream merge commits, and the associated upmerge policy with git merge, is under discussion, and if accepted, I will amend this documentation. 

I'll do the clarifications in this paragraph and will remove the "linear" term.

Reply:
@pabigot , changed to:
```
Changes to the main branch of a module repository, including synchronization
with upstream code base, can only be applied via pull requests. These pull
requests shall be *verifiable* by Zephyr CI and *mergable* (i.e. with Rebase
& Merge option, using Github UI). This ensures that the *downstream* module
repository history is incremental and always preserved (that is, it is never
re-written). This policy also allows to easily review the incoming changes
and run Zephyr CI, git lint, identity, and license checks directly on the set
of changes that are to be brought into the module repository.
```